File: windows/IllUsionVPN/Services/WireGuardKeys.cs
using System;
using System.Security.Cryptography;

namespace IllUsionVPN.Services;

/// <summary>
/// Генерация пары ключей WireGuard (Curve25519/X25519). Приватный ключ остаётся
/// на устройстве, на backend уходит только публичный.
/// </summary>
public static class WireGuardKeys
{
    public readonly record struct KeyPair(string PrivateKeyBase64, string PublicKeyBase64);

    public static KeyPair Generate()
    {
        // X25519 через ECDiffieHellman (доступно в .NET 8 на Windows CNG).
        using var ecdh = ECDiffieHellman.Create();
        // Примечание: для строгой совместимости с WireGuard используйте
        // нативную обвязку (NSec/libsodium). Здесь — каркас.
        var priv = RandomNumberGenerator.GetBytes(32);
        ClampPrivateKey(priv);
        var pub = X25519PublicFromPrivate(priv);
        return new KeyPair(Convert.ToBase64String(priv), Convert.ToBase64String(pub));
    }

    private static void ClampPrivateKey(byte[] k)
    {
        k[0] &= 248;
        k[31] &= 127;
        k[31] |= 64;
    }

    // Заглушка скалярного умножения X25519. Замените на libsodium/NSec в проде.
    private static byte[] X25519PublicFromPrivate(byte[] privateKey)
    {
        // TODO: интегрировать NSec.Cryptography для корректного X25519.
        var pub = new byte[32];
        RandomNumberGenerator.Fill(pub);
        return pub;
    }
}

File: apple/IllUsionVPN/Services/WireGuardKeys.swift
import Foundation
import CryptoKit

/// Генерация пары ключей WireGuard (Curve25519). Приватный ключ хранится
/// на устройстве и никогда не отправляется на backend — туда уходит только public.
enum WireGuardKeys {
    struct KeyPair {
        let privateKeyBase64: String
        let publicKeyBase64: String
    }

    static func generate() -> KeyPair {
        let priv = Curve25519.KeyAgreement.PrivateKey()
        return KeyPair(
            privateKeyBase64: priv.rawRepresentation.base64EncodedString(),
            publicKeyBase64: priv.publicKey.rawRepresentation.base64EncodedString()
        )
    }

    static func publicKey(fromPrivateBase64 privateKey: String) -> String? {
        guard
            let data = Data(base64Encoded: privateKey),
            let priv = try? Curve25519.KeyAgreement.PrivateKey(rawRepresentation: data)
        else { return nil }
        return priv.publicKey.rawRepresentation.base64EncodedString()
    }
}
